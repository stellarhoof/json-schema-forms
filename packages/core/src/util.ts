// import { getSchemaValue } from "@json-schema-forms/json-schema-utils"
// import {
//   Collection,
//   TraversalContext,
//   Tree,
// } from "@json-schema-forms/tree-utils"
// import _ from "lodash/fp.js"
//
// import { createField,Field } from "./createField.js"
//
// export const formTree = new Tree<{ children?: Collection<unknown> }>({
//   getChildren(field) {
//     return field.children as Collection<typeof field>
//   },
//   setChildren(field, children) {
//     field.children = children
//     return field
//   },
// })
//
// export const accumulate =
//   <T, R>(fn: (node: T, ctx: TraversalContext<T>) => R) =>
//   (acc: Record<string, R>, node: T, ctx: TraversalContext<T>) => {
//     const result = fn(node, ctx)
//     if (result) {
//       acc[ctx.path.join(".")] = result
//     }
//     return acc
//   }
//
// /**
//  * A simple check against the `schema.maxItems` json-schema keyword
//  *
//  * @see {@link "https://json-schema.org/draft/2020-12/json-schema-validation.html#name-maxitems"}
//  */
// export const canAddChild = <P extends object>(field?: Field<P>) => {
//   return (
//     field &&
//     field.schema.type === "array" &&
//     (field.schema.maxItems ?? -1) < field.value?.length &&
//     !field.disabled
//   )
// }
//
// /**
//  * Insert a new item at `index` to `items` and `value` arrays
//  *
//  * @remarks
//  *
//  * The new item will be cloned from `schema.items` and initialized just as a top-level form schema would (e.g. `field`s will be recursively added). An new, empty value will be generated by `getSchemaValue(schema.items)` and inserted into `value`.
//  *
//  * @param index - The index at which to insert the new item. Defaults to `value.length` if omitted.
//  */
// export const addChild = <P extends object>(field: Field<P>, index?: number) => {
//   if (field.schema.type === "array") {
//     index ??= _.size(field.value)
//     field.value.splice(index, 0, getSchemaValue(field.schema.items))
//     const created = createField({
//       id: Symbol(),
//       key: index,
//       parentId: field.id,
//       jsonSchema: field.schema.items,
//       context: field.context,
//     })
//     field.__state__.childrenIds.splice(index, 0, created.id)
//     for (const child of (field.children as any).slice(index + 1)) {
//       child.key = (child.key as number) + 1
//     }
//   }
// }
//
// /**
//  * A simple check against the `schema.minItems` json-schema keyword
//  *
//  * @see {@link "https://json-schema.org/draft/2020-12/json-schema-validation.html#name-minitems"}
//  */
// export const canRemoveChild = <P extends object>(field?: Field<P>) => {
//   return (
//     field &&
//     field.schema.type === "array" &&
//     (field.schema.minItems ?? Infinity) >= _.size(field.value) &&
//     !field.disabled
//   )
// }
//
// /**
//  * Remove an item at `index` from `items` and `value` arrays
//  *
//  * @param index - The index at which to remove the element. Defaults to `value.length` if omitted.
//  */
// export const removeChild = <P extends object>(
//   field: Field<P>,
//   index?: number,
// ) => {
//   if (field.schema.type === "array") {
//     index ??= _.size(field.value)
//     field.value.splice(index, 1)
//     const [id] = field.__state__.childrenIds.splice(index, 1)
//     const removed = field.context.fieldMap.get(id)!
//     removed.dispose()
//     for (const child of (field.children as any).slice(index)) {
//       child.key = (child.key as number) - 1
//     }
//   }
// }
