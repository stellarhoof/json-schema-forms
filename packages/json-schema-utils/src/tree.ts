import {
  Collection,
  TraversalContext,
  Tree,
} from "@json-schema-forms/tree-utils"
import _ from "lodash/fp.js"

export type TraversableSchema = {
  type: string
  properties?: Record<string, TraversableSchema>
  items?: TraversableSchema
}

export const jsonSchemaTree = new Tree<TraversableSchema>({
  getChildren(schema) {
    if (schema.type === "object" && schema.properties) {
      return schema.properties as Collection<typeof schema>
    }
    if (schema.type === "array" && schema.items) {
      return { "*": schema.items } as Record<
        string,
        TraversableSchema
      > as Collection<typeof schema>
    }
  },
  setChildren(schema, children) {
    if (schema.type === "object") {
      return {
        ...schema,
        properties: { ...schema.properties, ...children },
      } as typeof schema
    }
    if (schema.type === "array" && "*" in children) {
      return { ...schema, items: children["*"] } as typeof schema
    }
    return schema
  },
})

export const getChildrenSchemasFromValue =
  (value?: any) =>
  <T extends TraversableSchema>(
    schema: T,
    ctx: TraversalContext<T>
  ): Collection<T> | undefined => {
    if (schema.type === "object") {
      return schema.properties as Collection<T>
    }
    if (schema.type === "array") {
      return _.times(
        () => _.cloneDeep(schema.items),
        _.size(ctx.path ? _.get(ctx.path, value) : value)
      ) as Collection<T>
    }
  }
