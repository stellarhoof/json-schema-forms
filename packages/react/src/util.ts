import { HTMLInputTypeAttribute } from "react"
import { ReactField } from "./types.js"

/**
 * Input types that make sense for a form field
 */
type InputType = Exclude<
  HTMLInputTypeAttribute,
  | "button"
  | "image"
  | "reset"
  | "submit"
  // Superseded by `field.hidden`
  | "hidden"
  // This is some weird type union they've got :/
  | object
>

export const getFieldInputType = <P extends object>(
  field: ReactField<P>
): InputType => {
  const fromStringFormat: Record<string, InputType> = {
    date: "date",
    time: "time",
    "date-time": "datetime-local",
    email: "email",
    "idn-email": "email",
    uri: "url",
    iri: "url",
  }

  if (
    field.schema.type === "string" &&
    field.schema.format !== undefined &&
    field.schema.format in fromStringFormat
  ) {
    return fromStringFormat[field.schema.format]
  }

  if (
    (field.schema.type === "number" || field.schema.type === "integer") &&
    (field.schema.minimum ?? field.schema.exclusiveMinimum) !== undefined &&
    (field.schema.maximum ?? field.schema.exclusiveMaximum) !== undefined
  ) {
    return "range"
  }

  const fromSchemaType: Record<string, InputType> = {
    number: "number",
    integer: "number",
    boolean: "checkbox",
  }

  if (field.schema.type in fromSchemaType) {
    return fromSchemaType[field.schema.type]
  }

  return "text"
}
