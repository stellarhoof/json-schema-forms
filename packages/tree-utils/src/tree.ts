import type { Collection } from "./collection.js"
import * as collection from "./collection.js"

export interface TraversalContext<N, K = any> {
  key?: K
  path: K[]
  depth: number
  parents?: N[]
}

export interface ChildrenLens<T> {
  getChildren: <N extends T>(
    node: N,
    ctx: Omit<TraversalContext<N>, "children" | "parents">
  ) => Collection<N> | undefined
  setChildren: <N extends T>(
    node: N,
    children: Collection<N>,
    ctx: Omit<TraversalContext<N>, "children" | "parents">
  ) => N
}

const nextContext = <N>(
  parent: N,
  key: any,
  ctx: TraversalContext<N>
): TraversalContext<N> => ({
  key,
  path: key !== undefined ? [...ctx.path, key] : [],
  depth: ctx.depth + 1,
  parents: [parent, ...(ctx.parents ?? [])],
})

export class StopTraversal extends Error {}

export type Visitor<Fn> = Fn | { pre?: Fn; post?: Fn }

const getCbs = <Fn>(cb: Visitor<Fn>) =>
  (typeof cb === "function" ? { pre: cb, post: undefined } : cb) as {
    pre?: Fn
    post?: Fn
  }

export type ForEachCallback<N> = (node: N, ctx: TraversalContext<N>) => void

export type ForEachVisitor<N> = Visitor<ForEachCallback<N>>

export function forEach<T>(lens: ChildrenLens<T>) {
  return <N extends T>(cb: ForEachVisitor<N>, node: N): void => {
    const { pre, post } = getCbs(cb)
    const rec: ForEachCallback<N> = (node, ctx) => {
      const children = lens.getChildren(node, ctx)
      if (pre) {
        pre(node, ctx)
      }
      collection.forEach((value, key) => {
        rec(value, nextContext(node, key, ctx))
      }, children ?? [])
      if (post) {
        post(node, ctx)
      }
    }
    try {
      rec(node, { depth: 0, path: [] })
    } catch (e) {
      if (e instanceof StopTraversal) {
        return
      }
      throw e
    }
  }
}

export type ReduceCallback<N, R> = (
  result: R,
  node: N,
  ctx: TraversalContext<N>
) => R

export type ReduceVisitor<N, R> = Visitor<ReduceCallback<N, R>>

export function reduce<T>(lens: ChildrenLens<T>) {
  return <N extends T, R>(cb: ReduceVisitor<N, R>, result: R, node: N): R => {
    const { pre, post } = getCbs(cb)
    forEach(lens)(
      {
        pre: pre ? (...args) => (result = pre(result, ...args)) : undefined,
        post: post ? (...args) => (result = post(result, ...args)) : undefined,
      },
      node
    )
    return result
  }
}

export type FindCallback<N> = (node: N, ctx: TraversalContext<N>) => boolean

export type FindVisitor<N> = Visitor<FindCallback<N>>

export function find<T>(lens: ChildrenLens<T>) {
  return <N extends T>(pre: FindCallback<N>, node: N): void => {
    let result = undefined
    forEach(lens)(
      {
        pre: (node, ctx) => {
          if (pre(node, ctx)) {
            result = node
            throw new StopTraversal()
          }
        },
      },
      node
    )
    return result
  }
}

export type FilterCallback<N> = (node: N, ctx: TraversalContext<N>) => boolean

export type FilterVisitor<N> = Visitor<FilterCallback<N>>

export function filter<T>(lens: ChildrenLens<T>) {
  return <N extends T>(cb: FilterVisitor<N>, node: N): N | undefined => {
    const { pre, post } = getCbs(cb)

    const rec = (node: N, ctx: TraversalContext<N>) => {
      const children = lens.getChildren(node, ctx)

      if (pre && !pre(node, ctx)) {
        return
      }

      const mapped = children
        ? collection.map(
            (value, key) => rec(value, nextContext(node, key, ctx)),
            children
          )
        : undefined

      const filtered = mapped
        ? (collection.filter(
            (node) => node !== undefined,
            mapped
          ) as Collection<N>)
        : undefined

      node = filtered ? lens.setChildren(node, filtered, ctx) : node

      if (post && !post(node, ctx)) {
        return
      }

      return node
    }

    return rec(node, { depth: 0, path: [] })
  }
}

export type MapVisitor<N, R, Pre, Post> =
  | ((node: N, ctx: TraversalContext<R>) => R)
  | {
      pre?: Pre extends true ? (node: N, ctx: TraversalContext<R>) => R : never
      post?: Post extends true
        ? (
            node: Pre extends true ? R : N,
            ctx: TraversalContext<Pre extends true ? R : N>
          ) => R
        : never
    }

export function map<T>(lens: ChildrenLens<T>) {
  function map<N extends T, R extends T>(
    cb: MapVisitor<N, R, false, false>,
    node: N
  ): R
  function map<N extends T, R extends T>(
    cb: MapVisitor<N, R, true, false>,
    node: N
  ): R
  function map<N extends T, R extends T>(
    cb: MapVisitor<N, R, false, true>,
    node: N
  ): R
  function map<N extends T, R extends T>(
    cb: MapVisitor<N, R, true, true>,
    node: N
  ): R
  function map<N extends T, R extends T>(
    cb: MapVisitor<N, R, boolean, boolean>,
    node: N
  ): R {
    const { pre, post } = getCbs(cb)

    const rec = (node: N, ctx: TraversalContext<N | R>): R => {
      const children = lens.getChildren(node, ctx)

      let result = node as N | R

      if (pre) {
        result = pre(node, ctx as TraversalContext<R>)
      }

      const mapped = children
        ? collection.map(
            (value, key) => rec(value, nextContext(result, key, ctx)),
            children
          )
        : undefined

      if (post) {
        result = post(result as N, ctx as TraversalContext<R>)
      }

      return (mapped ? lens.setChildren(result, mapped, ctx) : result) as R
    }

    return rec(node, { depth: 0, path: [] })
  }

  return map
}
