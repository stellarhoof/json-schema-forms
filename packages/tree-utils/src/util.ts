export const isPlainObject = <V>(value: unknown): value is Record<any, V> =>
  !!value && typeof value === "object" && !Array.isArray(value)
