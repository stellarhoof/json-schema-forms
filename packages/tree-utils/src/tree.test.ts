import { describe, expect,it } from "vitest"

import { propLens,Tree } from "./index.js"
import { TraversalContext } from "./tree.js"
import { isPlainObject } from "./util.js"

type Json = string | number | boolean | null | Json[] | { [key: string]: Json }

const simpleTree = new Tree()

const simpleTreeValue: Json = {
  name: "The Empire Strikes Back",
  planets: ["Aldeeran", { name: "Tatooine" }],
}

const childrenTree = new Tree<{ children: any }>(propLens("children"))

const childrenTreeValue: Json = {
  name: "Anakin Skywalker",
  children: [
    {
      name: "Luke Skywalker",
      children: {
        film: { title: "A New Hope" },
      },
    },
  ],
}

describe("forEach", () => {
  type Acc = { node: Json; context: TraversalContext<Json> }[]

  const accumulateArgs =
    (acc: Acc) => (node: Json, context: TraversalContext<Json>) => {
      acc.push({ node, context })
    }

  it("pre simpleTree", () => {
    const args: Acc = []
    simpleTree.forEach({ pre: accumulateArgs(args) }, simpleTreeValue)
    expect(args).toMatchSnapshot()
  })

  it("post simpleTree", () => {
    const args: Acc = []
    simpleTree.forEach({ post: accumulateArgs(args) }, simpleTreeValue)
    expect(args).toMatchSnapshot()
  })

  it("pre childrenTree", () => {
    const args: Acc = []
    childrenTree.forEach({ pre: accumulateArgs(args) }, childrenTreeValue)
    expect(args).toMatchSnapshot()
  })

  it("post childrenTree", () => {
    const args: Acc = []
    childrenTree.forEach({ post: accumulateArgs(args) }, childrenTreeValue)
    expect(args).toMatchSnapshot()
  })
})

describe("reduce", () => {
  const init: { node: Json; ctx: TraversalContext<Json> }[] = []

  it("pre simpleTree", () => {
    const result = simpleTree.reduce(
      { pre: (acc, node, ctx) => [...acc, { node, ctx }] },
      init,
      simpleTreeValue
    )
    expect(result).toMatchSnapshot()
  })

  it("post simpleTree", () => {
    const result = simpleTree.reduce(
      { post: (acc, node, ctx) => [...acc, { node, ctx }] },
      init,
      simpleTreeValue
    )
    expect(result).toMatchSnapshot()
  })

  it("pre childrenTree", () => {
    const result = childrenTree.reduce(
      {
        pre: (acc, node, ctx) => [...acc, { node, ctx }],
      },
      init,
      childrenTreeValue
    )
    expect(result).toMatchSnapshot()
  })

  it("post childrenTree", () => {
    const result = childrenTree.reduce(
      {
        post: (acc, node, ctx) => [...acc, { node, ctx }],
      },
      init,
      childrenTreeValue
    )
    expect(result).toMatchSnapshot()
  })
})

describe("map", () => {
  const callback = (node: Json) => {
    if (typeof node === "string") {
      return `${node} added[string]`
    }
    if (Array.isArray(node)) {
      return [...node, "added[array]"]
    }
    if (isPlainObject(node)) {
      return { ...node, key: "added[object]" }
    }
    return node
  }

  it("pre simpleTree", () => {
    const result = simpleTree.map({ pre: callback }, simpleTreeValue)
    expect(result).toMatchSnapshot()
  })

  it("post simpleTree", () => {
    const result = simpleTree.map({ post: callback }, simpleTreeValue)
    expect(result).toMatchSnapshot()
  })

  it("pre childrenTree", () => {
    const result = childrenTree.map({ pre: callback }, childrenTreeValue)
    expect(result).toMatchSnapshot()
  })

  it("post childrenTree", () => {
    const result = childrenTree.map({ post: callback }, childrenTreeValue)
    expect(result).toMatchSnapshot()
  })
})

describe("filter", () => {
  it("should return empty when root node is filtered out", () => {
    expect(simpleTree.filter({ pre: (x) => !x }, "root")).toEqual(undefined)
    expect(simpleTree.filter({ pre: (x) => !x }, {})).toEqual(undefined)
  })

  it("should return root node when root node is not filtered out", () => {
    expect(simpleTree.filter({ pre: (x) => !!x }, "root")).toEqual("root")
    expect(simpleTree.filter({ pre: (x) => !!x }, {})).toEqual({})
  })

  it("should filter out nested collection before traversing into it", () => {
    const result = simpleTree.filter(
      { pre: (x) => !Array.isArray(x) },
      { a: "a", b: ["a", "b"] }
    )
    expect(result).toEqual({ a: "a" })
  })

  it("should filter out value inside nested collection", () => {
    const result = simpleTree.filter(
      { pre: (x: any) => x !== "b" },
      { a: "a", b: ["a", "b"] }
    )
    expect(result).toEqual({ a: "a", b: ["a"] })
  })

  it("should filter out empty values and collections in post-order traversal", () => {
    const result = simpleTree.filter(
      { post: (x: any) => x !== "b" && !(Array.isArray(x) && x.length === 0) },
      { a: "a", b: ["b", "b"] }
    )
    expect(result).toEqual({ a: "a" })
  })

  it("should filter out empty values and leave empty collections in pre-order traversal", () => {
    const result = simpleTree.filter(
      { pre: (x: any) => x !== "b" && !(Array.isArray(x) && x.length === 0) },
      { a: "a", b: ["b", "b"] }
    )
    expect(result).toEqual({ a: "a", b: [] })
  })
})

describe("find", () => {
  it("tree should return undefined given no node found", () => {
    const node = simpleTree.find(() => false, simpleTreeValue)
    expect(node).toEqual(undefined)
  })

  it("tree should return nested node/path given function predicate", () => {
    const node = simpleTree.find(
      (n) => typeof n === "string" && n === "Tatooine",
      simpleTreeValue
    )
    expect(node).toEqual("Tatooine")
  })

  it("childrenTree should return undefined given no node found", () => {
    const node = childrenTree.find(() => false, childrenTreeValue)
    expect(node).toEqual(undefined)
  })

  it("childrenTree should find nested node/path given function predicate", () => {
    const node = childrenTree.find(
      (n, { key }) => key === "film",
      childrenTreeValue
    )
    expect(node).toEqual({ title: "A New Hope" })
  })
})
