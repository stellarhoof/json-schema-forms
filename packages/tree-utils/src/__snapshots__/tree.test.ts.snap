// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`forEach > post childrenTree 1`] = `
[
  {
    "context": {
      "depth": 2,
      "key": "film",
      "parents": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
        "film",
      ],
    },
    "node": {
      "title": "A New Hope",
    },
  },
  {
    "context": {
      "depth": 1,
      "key": 0,
      "parents": [
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
      ],
    },
    "node": {
      "children": {
        "film": {
          "title": "A New Hope",
        },
      },
      "name": "Luke Skywalker",
    },
  },
  {
    "context": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "children": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
      ],
      "name": "Anakin Skywalker",
    },
  },
]
`;

exports[`forEach > post simpleTree 1`] = `
[
  {
    "context": {
      "depth": 1,
      "key": "name",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "name",
      ],
    },
    "node": "The Empire Strikes Back",
  },
  {
    "context": {
      "depth": 2,
      "key": 0,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        0,
      ],
    },
    "node": "Aldeeran",
  },
  {
    "context": {
      "depth": 3,
      "key": "name",
      "parents": [
        {
          "name": "Tatooine",
        },
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
        "name",
      ],
    },
    "node": "Tatooine",
  },
  {
    "context": {
      "depth": 2,
      "key": 1,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
      ],
    },
    "node": {
      "name": "Tatooine",
    },
  },
  {
    "context": {
      "depth": 1,
      "key": "planets",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
      ],
    },
    "node": [
      "Aldeeran",
      {
        "name": "Tatooine",
      },
    ],
  },
  {
    "context": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "name": "The Empire Strikes Back",
      "planets": [
        "Aldeeran",
        {
          "name": "Tatooine",
        },
      ],
    },
  },
]
`;

exports[`forEach > pre childrenTree 1`] = `
[
  {
    "context": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "children": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
      ],
      "name": "Anakin Skywalker",
    },
  },
  {
    "context": {
      "depth": 1,
      "key": 0,
      "parents": [
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
      ],
    },
    "node": {
      "children": {
        "film": {
          "title": "A New Hope",
        },
      },
      "name": "Luke Skywalker",
    },
  },
  {
    "context": {
      "depth": 2,
      "key": "film",
      "parents": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
        "film",
      ],
    },
    "node": {
      "title": "A New Hope",
    },
  },
]
`;

exports[`forEach > pre simpleTree 1`] = `
[
  {
    "context": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "name": "The Empire Strikes Back",
      "planets": [
        "Aldeeran",
        {
          "name": "Tatooine",
        },
      ],
    },
  },
  {
    "context": {
      "depth": 1,
      "key": "name",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "name",
      ],
    },
    "node": "The Empire Strikes Back",
  },
  {
    "context": {
      "depth": 1,
      "key": "planets",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
      ],
    },
    "node": [
      "Aldeeran",
      {
        "name": "Tatooine",
      },
    ],
  },
  {
    "context": {
      "depth": 2,
      "key": 0,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        0,
      ],
    },
    "node": "Aldeeran",
  },
  {
    "context": {
      "depth": 2,
      "key": 1,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
      ],
    },
    "node": {
      "name": "Tatooine",
    },
  },
  {
    "context": {
      "depth": 3,
      "key": "name",
      "parents": [
        {
          "name": "Tatooine",
        },
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
        "name",
      ],
    },
    "node": "Tatooine",
  },
]
`;

exports[`map > post childrenTree 1`] = `
{
  "children": [
    {
      "children": {
        "film": {
          "key": "added[object]",
          "title": "A New Hope",
        },
      },
      "key": "added[object]",
      "name": "Luke Skywalker",
    },
  ],
  "key": "added[object]",
  "name": "Anakin Skywalker",
}
`;

exports[`map > post simpleTree 1`] = `
{
  "name": "The Empire Strikes Back added[string]",
  "planets": [
    "Aldeeran added[string]",
    {
      "name": "Tatooine added[string]",
    },
  ],
}
`;

exports[`map > pre childrenTree 1`] = `
{
  "children": [
    {
      "children": {
        "film": {
          "key": "added[object]",
          "title": "A New Hope",
        },
      },
      "key": "added[object]",
      "name": "Luke Skywalker",
    },
  ],
  "key": "added[object]",
  "name": "Anakin Skywalker",
}
`;

exports[`map > pre simpleTree 1`] = `
{
  "name": "The Empire Strikes Back added[string]",
  "planets": [
    "Aldeeran added[string]",
    {
      "name": "Tatooine added[string]",
    },
  ],
}
`;

exports[`reduce > post childrenTree 1`] = `
[
  {
    "ctx": {
      "depth": 2,
      "key": "film",
      "parents": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
        "film",
      ],
    },
    "node": {
      "title": "A New Hope",
    },
  },
  {
    "ctx": {
      "depth": 1,
      "key": 0,
      "parents": [
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
      ],
    },
    "node": {
      "children": {
        "film": {
          "title": "A New Hope",
        },
      },
      "name": "Luke Skywalker",
    },
  },
  {
    "ctx": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "children": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
      ],
      "name": "Anakin Skywalker",
    },
  },
]
`;

exports[`reduce > post simpleTree 1`] = `
[
  {
    "ctx": {
      "depth": 1,
      "key": "name",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "name",
      ],
    },
    "node": "The Empire Strikes Back",
  },
  {
    "ctx": {
      "depth": 2,
      "key": 0,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        0,
      ],
    },
    "node": "Aldeeran",
  },
  {
    "ctx": {
      "depth": 3,
      "key": "name",
      "parents": [
        {
          "name": "Tatooine",
        },
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
        "name",
      ],
    },
    "node": "Tatooine",
  },
  {
    "ctx": {
      "depth": 2,
      "key": 1,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
      ],
    },
    "node": {
      "name": "Tatooine",
    },
  },
  {
    "ctx": {
      "depth": 1,
      "key": "planets",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
      ],
    },
    "node": [
      "Aldeeran",
      {
        "name": "Tatooine",
      },
    ],
  },
  {
    "ctx": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "name": "The Empire Strikes Back",
      "planets": [
        "Aldeeran",
        {
          "name": "Tatooine",
        },
      ],
    },
  },
]
`;

exports[`reduce > pre childrenTree 1`] = `
[
  {
    "ctx": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "children": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
      ],
      "name": "Anakin Skywalker",
    },
  },
  {
    "ctx": {
      "depth": 1,
      "key": 0,
      "parents": [
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
      ],
    },
    "node": {
      "children": {
        "film": {
          "title": "A New Hope",
        },
      },
      "name": "Luke Skywalker",
    },
  },
  {
    "ctx": {
      "depth": 2,
      "key": "film",
      "parents": [
        {
          "children": {
            "film": {
              "title": "A New Hope",
            },
          },
          "name": "Luke Skywalker",
        },
        {
          "children": [
            {
              "children": {
                "film": {
                  "title": "A New Hope",
                },
              },
              "name": "Luke Skywalker",
            },
          ],
          "name": "Anakin Skywalker",
        },
      ],
      "path": [
        0,
        "film",
      ],
    },
    "node": {
      "title": "A New Hope",
    },
  },
]
`;

exports[`reduce > pre simpleTree 1`] = `
[
  {
    "ctx": {
      "depth": 0,
      "path": [],
    },
    "node": {
      "name": "The Empire Strikes Back",
      "planets": [
        "Aldeeran",
        {
          "name": "Tatooine",
        },
      ],
    },
  },
  {
    "ctx": {
      "depth": 1,
      "key": "name",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "name",
      ],
    },
    "node": "The Empire Strikes Back",
  },
  {
    "ctx": {
      "depth": 1,
      "key": "planets",
      "parents": [
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
      ],
    },
    "node": [
      "Aldeeran",
      {
        "name": "Tatooine",
      },
    ],
  },
  {
    "ctx": {
      "depth": 2,
      "key": 0,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        0,
      ],
    },
    "node": "Aldeeran",
  },
  {
    "ctx": {
      "depth": 2,
      "key": 1,
      "parents": [
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
      ],
    },
    "node": {
      "name": "Tatooine",
    },
  },
  {
    "ctx": {
      "depth": 3,
      "key": "name",
      "parents": [
        {
          "name": "Tatooine",
        },
        [
          "Aldeeran",
          {
            "name": "Tatooine",
          },
        ],
        {
          "name": "The Empire Strikes Back",
          "planets": [
            "Aldeeran",
            {
              "name": "Tatooine",
            },
          ],
        },
      ],
      "path": [
        "planets",
        1,
        "name",
      ],
    },
    "node": "Tatooine",
  },
]
`;
